src/algebra/algebra/unitization.lean

```diff
@@ -0,0 +1,498 @@
+/-
+Copyright (c) 2022 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+
+import algebra.algebra.basic
+import linear_algebra.prod
+import algebra.non_unital_alg_hom
+
+/-!
+# Unitization of a non-unital algebra
+
+Given a non-unital `R`-algebra `A` (given via the type classes
+`[non_unital_ring A] [module R A] [smul_comm_class R A A] [is_scalar_tower R A A]`) we construct
+the minimal unital `R`-algebra containing `A` as an ideal. This object `algebra.unitization R A` is
+a type synonym for `R × A` on which we place a different multiplicative structure, namely,
+`(r₁, a₁) * (r₂, a₂) = (r₁ * r₂, r₁ • a₂ + r₂ • a₁ + a₁ * a₂)` where the multiplicative identity
+is `(1, 0)`.
+
+Note, when `A` is a *unital* `R`-algebra, then `unitization R A` constructs a new multiplicative
+identity different from the old one, and so in general `unitization R A` and `A` will not be
+isomorphic even in the unital case. This approach actually has nice functorial properties.
+
+There is a natural coercion from `A` to `unitization R A` given by `λ a, (0, a)`, the image
+of which is a proper ideal (TODO), and when `R` is a field this ideal is maximal. Moreover,
+this ideal is always an essential ideal (it has nontrivial intersection with every other nontrivial
+ideal).
+
+Every non-unital algebra homomorphism from `A` into a *unital* `R`-algebra `B` has a unique
+extension to a (unital) algebra homomorphism from `unitization R A` to `B`.
+
+## Main definitions
+
+* `unitization R A`: the unitization of a non-unital `R`-algebra `A`.
+* `unitization.algebra`: the unitization of `A` as a (unital) `R`-algebra.
+* `unitization.coe_non_unital_alg_hom`: coercion as a non-unital algebra homomorphism.
+* `non_unital_alg_hom.to_alg_hom φ`: the extension of a non-unital algebra homomorphism `φ : A → B`
+  into a unital `R`-algebra `B` to an algebra homomorphism `unitization R A →ₐ[R] B`.
+
+## Main results
+
+* `non_unital_alg_hom.to_alg_hom_unique`: the extension is unique
+
+## TODO
+
+* prove the unitization operation is a functor between the appropriate categories
+* prove the image of the coercion is an essential ideal, maximal if scalars are a field.
+-/
+
+/-- The minimal unitization of a non-unital `R`-algebra `A`. This is just a type synonym for
+`R × A`.-/
+def unitization (R A : Type*) := R × A
+
+namespace unitization
+
+section basic
+
+variables {R A : Type*}
+
+/-- The canonical inclusion `R → unitization R A`. -/
+def inl [has_zero A] (r : R) : unitization R A :=
+(r, 0)
+
+/-- The canonical inclusion `A → unitization R A`. -/
+instance [has_zero R] : has_coe_t A (unitization R A) := { coe := λ a, (0, a) }
+
+/-- The canonical projection `unitization R A → R`. -/
+def fst (x : unitization R A) : R :=
+x.1
+
+/-- The canonical projection `unitization R A → A`. -/
+def snd (x : unitization R A) : A :=
+x.2
```

Comment:
Do we actually need these? If we remove them it breaks a few lines below, but just replacing them by `prod.fst` and `prod.snd` works.

Reply:
You're correct that we don't strictly need them, but I think we definitely want them. If we use `prod.fst` and `prod.snd` then Lean will be continually re-interpreting things of type `unitization R A` as terms of type `R × A`. This is definitely not what we want, and, just to hazard a guess, it would probably make the `simp_nf` linter fail all over the place (because then a whole bunch of simp lemmas in this file would be about `prod` instead, and they already exist for `prod`; you can see this in the widget panel if you make your suggested change and then inspect a simp lemma involving `fst`).

Reply:
Fair enough!